src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1214,4 +1233,45 @@ protected Engine newEngine() {
     public boolean allowsPrimaryPromotion() {
         return true;
     }
+
+    // pkg private for testing
+    void persistMetadata(ShardRouting newRouting, ShardStateMetaData oldShardStateMetadata) {
```

Comment:
imho, we should just always persist the newRouting. The oldShardStateMetadata just complicates things and as far as I can tell doesn't add much value - we call this function very rarely.